src/database/operations/assert/assertProposalAuthorization.ts

```diff
@@ -0,0 +1,17 @@
+import { db } from '../../db';
+import { NotFoundError } from '../../../errors';
+import type { CheckResult } from '../../../types';
+
+export const assertProposalAuthorization = async (
+	id: number,
+	userId: number,
+): Promise<void> => {
+	const result = await db.sql<CheckResult>('proposals.checkAuthorization', {
+		id,
+		userId,
+	});
+
+	if (result.rows[0]?.result !== true) {
+		throw new NotFoundError(`The proposal was not found (id: ${id})`);
```

Comment:
This error looks incorrect, but we want to match the "not found" response from a security PoV.  We may want to refactor this so that the middleware is what smooths this out.

For instance: Maybe `NotFoundError` should be changed to `NotFoundOrNoAccessError` and the middleware should strip out the message.

Reply:
Caveat: I haven't tried it yet locally.

I think public (unauthenticated) users should already get a 401 on every attempt to access any proposal whether it exists or not. So unauthenticated users shouldn't be able to discover whether a proposal ID exists. I think if a user is already authenticated but does not have access to a particular proposal, the clarity given by a 403 instead of a 404 outweighs the risks posed by an already-authenticated user discovering there exists or does not exist a proposal for a given ID. In other words, I would expect 401 for all attempts for public users, 404 for authenticated users when the ID truly does not exist, and 403 for authenticated users who are not permitted to access the resource.

At the same time, I understand the reasoning and it is a judgment call.

Reply:
Hmm, I guess when doing a pre-flight query this way, there is no way to tell if it is due to not being found or not having access, interesting.

Reply:
So my thinking here is: do we want the user to know a resource exists that they don't have access to?  IMHO even if they have an account the answer is no, but that's a generic / conservative opinion and if we think folks should know about proposals they can't read then that's OK too.

If we want 403 instead of 404 for items that exist with no access I can modify the preflight check so that instead of returning a boolean it returns something to indicate "no access" vs "doesn't exist"

Reply:
I think if you're authenticated to the PDC you have broad read access and narrow write access. I do want the 403 for authenticated users that do not have authorization. I am also assuming that proposals can be read almost all the time by almost all PDC users and that the exception is to deny access. I imagine someone on a call saying "let's look at proposal 944" and if one person gets a "not found" while the other sees the proposal that would be very confusing. Either it exists or doesn't. `NotFoundOrUnauthorizedError` would also make it a bit clearer that it could be either. I am curious what others think, though.

Reply:
Ah I think there may be a mismatch / things shifted when you weren't on call -- we're adding these checks in order to limit access to most proposals for most users right now (though I don't think that "this is an edge case" should necessarily drive whether 404 or 403 is the right call -- if we want to let people know information about data they can't access then that probably wouldn't depend on how many people fit into that category).

What about for other kinds of lookup / natural keys?  E.g. what if someone is looking up an organization by EIN and it exists but they don't have access to that fact?  It would be nice to have predictable / consistent behavior that will handle a variety of cases.

FWIW: my understanding is that it's normal for an API to either ALWAYS use 404 or ALWAYS use 403 for resource access in order to avoid sharing information about entity state to unauthorized users.  Not saying we have to do it, but just wanted to make sure it's clear that it isn't a new concept / something that we're inventing.

For your example would it be less confusing if we always returned 403 rather than 404?

Reply:
My comments assumed we were in a largely "the purpose of PDC is data sharing among PDC folks" mode. If that has changed, then yeah, I think being more opaque and restrictive makes sense.

Reply:
> What about for other kinds of lookup / natural keys? E.g. what if someone is looking up an organization by EIN and it exists but they don't have access to that fact?

I would lean toward "no, that's not a mostly-meaningless auto-generated ID so limit access there", but again, if we were in a PDC is for data sharing among PDC folks mode, I would be puzzled that authenticated users couldn't see that. The ground has shifted, perhaps.

Reply:
> For your example would it be less confusing if we always returned 403 rather than 404?

I think the least confusing is distinguishing not found from not authorized but yes, returning 403 every time is better than returning 404 every time.